tests/substrate.rs

```diff
@@ -1,1145 +1,866 @@
+        Ok(vm.accounts.len() as u64)
+    }
 
-                set_seal_value!("seal_address", dest_ptr, len_ptr, &scratch);
+    #[seal(0)]
+    fn seal_minimum_balance(out_ptr: u32, out_len_ptr: u32) -> Result<(), Trap> {
+        assert!(read_len(mem, out_len_ptr) >= 16);
+        write_buf(mem, out_ptr, &500u128.to_le_bytes());
+        Ok(())
+    }
 
-                Ok(None)
-            }
-            Some(SubstrateExternal::seal_caller) => {
-                let dest_ptr: u32 = args.nth_checked(0)?;
+    #[seal(1)]
+    fn seal_instantiate(
+        code_hash_ptr: u32,
+        _gas: u64,
+        value_ptr: u32,
+        input_data_ptr: u32,
+        input_data_len: u32,
+        address_ptr: u32,
+        address_len_ptr: u32,
+        output_ptr: u32,
+        output_len_ptr: u32,
+        salt_ptr: u32,
+        salt_len: u32,
+    ) -> Result<u32, Trap> {
+        let code_hash = read_account(mem, code_hash_ptr);
+        let salt = read_buf(mem, salt_ptr, salt_len);
+        let input = read_buf(mem, input_data_ptr, input_data_len);
+        let value = read_value(mem, value_ptr);
+
+        if value > vm.accounts[vm.account].value {
+            return Ok(5); // ReturnCode::TransferFailed
+        }
 
-                let len_ptr: u32 = args.nth_checked(1)?;
-                let scratch = self.vm.caller;
+        let ((flags, data), state) = match vm.deploy(code_hash, value, &salt, input) {
+            Some(Ok(state)) => ((state.data().output.as_data()), state),
+            Some(Err(_)) => return Ok(1), // ReturnCode::CalleeTrapped
+            None => return Ok(7),         // ReturnCode::CodeNotFound
+        };
 
-                set_seal_value!("seal_caller", dest_ptr, len_ptr, &scratch);
+        if output_len_ptr != u32::MAX {
+            write_buf(mem, output_ptr, &data);
+            write_buf(mem, output_len_ptr, &(data.len() as u32).to_le_bytes());
+        }
 
-                Ok(None)
-            }
-            Some(SubstrateExternal::seal_balance) => {
-                let dest_ptr: u32 = args.nth_checked(0)?;
-                let len_ptr: u32 = args.nth_checked(1)?;
+        let address = state.data().accounts.last().unwrap().address;
+        write_buf(mem, address_ptr, &address);
+        write_buf(mem, address_len_ptr, &(address.len() as u32).to_le_bytes());
 
-                let scratch = self.accounts[&self.vm.account].1.to_le_bytes();
+        if flags == 2 {
+            return Ok(2); // ReturnCode::CalleeReverted
+        }
 
-                set_seal_value!("seal_balance", dest_ptr, len_ptr, &scratch);
+        vm.accept_state(state.into_data(), value);
+        Ok(0)
+    }
 
-                Ok(None)
-            }
-            Some(SubstrateExternal::seal_minimum_balance) => {
-                let dest_ptr: u32 = args.nth_checked(0)?;
-                let len_ptr: u32 = args.nth_checked(1)?;
+    #[seal(0)]
+    fn seal_transfer(account_ptr: u32, _: u32, value_ptr: u32, _: u32) -> Result<u32, Trap> {
```

Comment:
`seal_tranfer` no longer cheecks that the 2nd and 5th argument are correct; we should ensure that our  code sets the correct length, else it will not work on a real node.